File: fsharp/src/Parsers.fs
module Parsers

open FSharp.Data
open System

type Sport =
    | Kiteboarding
    | SUP
    | Skiing
    | Snowboarding
    | MountainBiking
    | Hiking
    | Running
    | Paragliding
    | Surfing
    | WingFoiling
    | Parawinging

type SeshType =
    | Spot
    | Downwinder
    | Roundwinder

type BoardType =
    | Twintip
    | Hydrofoil
    | Surfboard
    | SUP
    | Skis
    | Snowboard
    | Other

type Normalized =
    { Date: DateTime
      Sport: Sport
      Hours: double
      WindAvg: int option
      WindGust: int option
      KiteSize: string list option
      WingSize: string list option
      SeshType: SeshType option
      BoardType: BoardType list option
      Foil: string list option
      Board: string option
      Location: string option
      Comments: string option }

let private maybeString (s: string) : string option =
    match s.Trim() with
    | "" -> None
    | s -> Some s

let private parseInt = maybeString >> Option.map int

let private parseMany (s: string) =
    let split (s: string) = s.Split "," |> Array.toList
    let trim (s: string) = s.Trim()

    let items =
        s
        |> maybeString
        |> Option.map (split >> List.map trim)
        |> Option.map (List.map maybeString >> List.choose id)

    match items with
    | None -> None
    | Some [] -> None
    | Some xs -> Some xs

let private normalizeSport (sport: string) =
    match sport.Trim() with
    | "Kiteboarding" -> Kiteboarding
    | "SUP" -> Sport.SUP
    | "Skiing" -> Skiing
    | "Snowboarding" -> Snowboarding
    | "Mountain Biking" -> MountainBiking
    | "Hiking" -> Hiking
    | "Running" -> Running
    | "Paragliding" -> Paragliding
    | "Surfing" -> Surfing
    | "Surf" -> Surfing
    | "Wing foiling" -> WingFoiling
    | "Parawinging" -> Parawinging
    | other -> failwith $"Unhandled sport: `{other}`"

let private normalizeSeshType =
    function
    | Some "Spot" -> Some Spot
    | Some "Downwinder" -> Some Downwinder
    | Some "Roundwinder" -> Some Roundwinder
    | _ -> None

let private normalizeBoardType =
    function
    | "Twintip"
    | "Twintp" -> Twintip // Fix typo
    | "Hydrofoil" -> Hydrofoil
    | "Surfboard"
    | "Strapless" -> Surfboard
    | "SUP" -> SUP
    | "Skis" -> Skis
    | "Snowboard" -> Snowboard
    | "Skim" -> Other
    | other -> failwith $"Unhandled board type: `{other}`"

let private parseBoardTypes s =
    s |> parseMany |> Option.map (List.map normalizeBoardType)

// Some kite sizes are defined as "12m" and others just "12"
let private normalizeKiteSize (kites: string list option) =
    kites |> Option.map (List.map (fun kite -> kite.Replace("m", "")))

let private parse2012 (row: CsvRow) : Normalized =
    { Date = row.GetColumn "Date" |> DateTime.Parse
      Sport = row.GetColumn "Sport" |> normalizeSport
      Hours = row.GetColumn "Hours" |> double
      WindAvg = row.GetColumn "Lull (kn)" |> parseInt
      WindGust = row.GetColumn "Gust (kn)" |> parseInt
      KiteSize = row.GetColumn "Kite Size" |> parseMany |> normalizeKiteSize
      WingSize = None
      SeshType = row.GetColumn "Type" |> maybeString |> normalizeSeshType
      BoardType = None
      Foil = None
      Board = None
      Location = None
      Comments = row.GetColumn "Comments" |> maybeString }

let private parse2013 (row: CsvRow) : Normalized =
    { Date = row.GetColumn "Date" |> DateTime.Parse
      Sport = row.GetColumn "Sport" |> normalizeSport
      Hours = row.GetColumn "Hours" |> double
      WindAvg = row.GetColumn "Lull" |> parseInt
      WindGust = row.GetColumn "Gust" |> parseInt
      KiteSize = row.GetColumn "Kite" |> parseMany |> normalizeKiteSize
      WingSize = None
      SeshType = row.GetColumn "Type" |> maybeString |> normalizeSeshType
      BoardType = None
      Foil = None
      Board = None
      Location = None
      Comments = row.GetColumn "Comments" |> maybeString }

let private parse2014 (row: CsvRow) : Normalized =
    { Date = row.GetColumn "Day" |> DateTime.Parse
      Sport = row.GetColumn "Sport" |> normalizeSport
      Hours = row.GetColumn "Hours" |> double
      WindAvg = row.GetColumn "Lull (kn)" |> parseInt
      WindGust = row.GetColumn "Gust (kn)" |> parseInt
      KiteSize = row.GetColumn "Kite Size" |> parseMany |> normalizeKiteSize
      WingSize = None
      SeshType = row.GetColumn "Type" |> maybeString |> normalizeSeshType
      BoardType = None
      Foil = None
      Board = None
      Location = row.GetColumn "Location" |> maybeString
      Comments = row.GetColumn "Comments" |> maybeString }

let private parse2015 (row: CsvRow) : Normalized =
    { Date = row.GetColumn "Date" |> DateTime.Parse
      Sport = row.GetColumn "Sport" |> normalizeSport
      Hours = row.GetColumn "Hours" |> double
      WindAvg = row.GetColumn "Lull" |> parseInt
      WindGust = row.GetColumn "Gust" |> parseInt
      KiteSize = row.GetColumn "Kite" |> parseMany |> normalizeKiteSize
      WingSize = None
      SeshType = row.GetColumn "Type" |> maybeString |> normalizeSeshType
      BoardType = None
      Foil = None
      Board = None
      Location = row.GetColumn "Location" |> maybeString
      Comments = row.GetColumn "Comments" |> maybeString }

let private parse2016 (row: CsvRow) : Normalized =
    { Date = row.GetColumn "Date" |> DateTime.Parse
      Sport = row.GetColumn "Sport" |> normalizeSport
      Hours = row.GetColumn "Hours" |> double
      WindAvg = row.GetColumn "Lull (kts)" |> parseInt
      WindGust = row.GetColumn "Gust (kts)" |> parseInt
      KiteSize = row.GetColumn "Kite" |> parseMany |> normalizeKiteSize
      WingSize = None
      SeshType = row.GetColumn "Type" |> maybeString |> normalizeSeshType
      BoardType = row.GetColumn "Board" |> parseBoardTypes
      Foil = None
      Board = None
      Location = row.GetColumn "Location" |> maybeString
      Comments = row.GetColumn "Comments" |> maybeString }

let private parse2022 (row: CsvRow) : Normalized =
    { Date = row.GetColumn "Date" |> DateTime.Parse
      Sport = row.GetColumn "Sport" |> normalizeSport
      Hours = row.GetColumn "Hours" |> double
      WindAvg = row.GetColumn "Avg (kts)" |> parseInt
      WindGust = row.GetColumn "Gust (kts)" |> parseInt
      KiteSize = row.GetColumn "Kite" |> parseMany |> normalizeKiteSize
      WingSize = row.GetColumn "Wing" |> parseMany
      SeshType = row.GetColumn "Type" |> maybeString |> normalizeSeshType
      BoardType = row.GetColumn "Board Type" |> parseBoardTypes
      Foil = row.GetColumn "Foil" |> parseMany
      Board = row.GetColumn "Foil Board" |> maybeString
      Location = row.GetColumn "Location" |> maybeString
      Comments = row.GetColumn "Comments" |> maybeString }

let private parse2024 (row: CsvRow) : Normalized =
    { Date = row.GetColumn "Date" |> DateTime.Parse
      Sport = row.GetColumn "Sport" |> normalizeSport
      Hours = row.GetColumn "Hours" |> double
      WindAvg = row.GetColumn "Avg (kts)" |> parseInt
      WindGust = row.GetColumn "Gust (kts)" |> parseInt
      KiteSize = row.GetColumn "Kite" |> parseMany |> normalizeKiteSize
      WingSize = row.GetColumn "Wing" |> parseMany
      SeshType = row.GetColumn "Type" |> maybeString |> normalizeSeshType
      BoardType = row.GetColumn "Board Type" |> parseBoardTypes
      Foil = row.GetColumn "Foil" |> parseMany
      Board = row.GetColumn "Foil Board" |> maybeString
      Location = row.GetColumn "Location" |> maybeString
      Comments = row.GetColumn "Comments" |> maybeString }

let private parse2025 (row: CsvRow) : Normalized =
    { Date = row.GetColumn "Date" |> DateTime.Parse
      Sport = row.GetColumn "Sport" |> normalizeSport
      Hours = row.GetColumn "Hours" |> double
      WindAvg = row.GetColumn "Avg (kts)" |> parseInt
      WindGust = row.GetColumn "Gust (kts)" |> parseInt
      KiteSize = row.GetColumn "Kite" |> parseMany |> normalizeKiteSize
      WingSize = row.GetColumn "Wing" |> parseMany
      SeshType = row.GetColumn "Type" |> maybeString |> normalizeSeshType
      BoardType = row.GetColumn "Board Type" |> parseBoardTypes
      Foil = row.GetColumn "Foil" |> parseMany
      Board = row.GetColumn "Board" |> maybeString
      Location = row.GetColumn "Location" |> maybeString
      Comments = row.GetColumn "Comments" |> maybeString }

let parseFile (schema: int) (path: string) =
    CsvFile.Load(path).Rows
    |> Seq.map (
        match schema with
        | 2012 -> parse2012
        | 2013 -> parse2013
        | 2014 -> parse2014
        | 2015 -> parse2015
        | 2016
        | 2017
        | 2018
        | 2019
        | 2020
        | 2021 -> parse2016
        | 2022
        | 2023 -> parse2022
        | 2024 -> parse2024
        | 2025
        | 2026 -> parse2025
        | other -> failwith $"Schema for `{other}` not defined yet!"
    )
    |> Seq.toList

File: ocaml/lib/parsers.ml
open Types

let maybe_string s =
  match String.trim s with
  | "" -> None
  | s -> Some s
;;

let parse_int s = maybe_string s |> Option.map int_of_string

let parse_many s =
  let split s = String.split_on_char ',' s in
  let items =
    maybe_string s |> Option.map (fun s -> split s |> List.map String.trim) |> Option.map (List.filter_map maybe_string)
  in
  match items with
  | None -> None
  | Some [] -> None
  | Some xs -> Some xs
;;

let normalize_sport sport =
  match String.trim sport with
  | "Kiteboarding" -> Kiteboarding
  | "SUP" -> SUP
  | "Skiing" -> Skiing
  | "Snowboarding" -> Snowboarding
  | "Mountain Biking" -> MountainBiking
  | "Hiking" -> Hiking
  | "Running" -> Running
  | "Paragliding" -> Paragliding
  | "Surfing" | "Surf" -> Surfing
  | "Wing foiling" -> WingFoiling
  | "Parawinging" -> Parawinging
  | other -> failwith (Printf.sprintf "Unhandled sport: `%s`" other)
;;

let normalize_sesh_type = function
  | Some "Spot" -> Some Spot
  | Some "Downwinder" -> Some Downwinder
  | Some "Roundwinder" -> Some Roundwinder
  | _ -> None
;;

let normalize_board_type = function
  | "Twintip" | "Twintp" -> Twintip (* Fix typo *)
  | "Hydrofoil" -> Hydrofoil
  | "Surfboard" | "Strapless" -> Surfboard
  | "SUP" -> SUPBoard
  | "Skis" -> Skis
  | "Snowboard" -> Snowboard
  | "Skim" -> Other
  | other -> failwith (Printf.sprintf "Unhandled board type: `%s`" other)
;;

let parse_board_types s = parse_many s |> Option.map (List.map normalize_board_type)

(* Some kite sizes are defined as "12m" and others just "12" *)
let normalize_kite_size kites =
  let digits_only s = Str.global_replace (Str.regexp "[^0-9]") "" s in
  kites |> Option.map (List.map digits_only)
;;

(* Parse date from M/D/YYYY to YYYY-MM-DD *)
let parse_date s = Scanf.sscanf s "%d/%d/%d" (fun m d y -> Printf.sprintf "%04d-%02d-%02d" y m d)

let parse_2012 row =
  match row with
  | [ date; sport; hours; lull; gust; kite_size; sesh_type; comments ] ->
    { date = parse_date date
    ; sport = normalize_sport sport
    ; hours = float_of_string hours
    ; wind_avg = parse_int lull
    ; wind_gust = parse_int gust
    ; kite_size = parse_many kite_size |> normalize_kite_size
    ; wing_size = None
    ; sesh_type = maybe_string sesh_type |> normalize_sesh_type
    ; board_type = None
    ; foil = None
    ; board = None
    ; location = None
    ; comments = maybe_string comments
    }
  | _ -> failwith "parse_2012: unexpected row format"
;;

let parse_2013 row =
  match row with
  | [ date; sport; hours; lull; gust; kite; sesh_type; comments ] ->
    { date = parse_date date
    ; sport = normalize_sport sport
    ; hours = float_of_string hours
    ; wind_avg = parse_int lull
    ; wind_gust = parse_int gust
    ; kite_size = parse_many kite |> normalize_kite_size
    ; wing_size = None
    ; sesh_type = maybe_string sesh_type |> normalize_sesh_type
    ; board_type = None
    ; foil = None
    ; board = None
    ; location = None
    ; comments = maybe_string comments
    }
  | _ -> failwith "parse_2013: unexpected row format"
;;

let parse_2015 row =
  match row with
  (* Some rows have an empty trailing column *)
  | [ date; sport; hours; lull; gust; kite; sesh_type; location; comments ]
  | [ date; sport; hours; lull; gust; kite; sesh_type; location; comments; _ ] ->
    { date = parse_date date
    ; sport = normalize_sport sport
    ; hours = float_of_string hours
    ; wind_avg = parse_int lull
    ; wind_gust = parse_int gust
    ; kite_size = parse_many kite |> normalize_kite_size
    ; wing_size = None
    ; sesh_type = maybe_string sesh_type |> normalize_sesh_type
    ; board_type = None
    ; foil = None
    ; board = None
    ; location = maybe_string location
    ; comments = maybe_string comments
    }
  | _ -> failwith "parse_2015: unexpected row format"
;;

let parse_2014 row =
  match row with
  | [ date; sport; hours; lull; gust; kite_size; sesh_type; location; comments ]
  | [ date; sport; hours; lull; gust; kite_size; sesh_type; location; comments; _ ] ->
    { date = parse_date date
    ; sport = normalize_sport sport
    ; hours = float_of_string hours
    ; wind_avg = parse_int lull
    ; wind_gust = parse_int gust
    ; kite_size = parse_many kite_size |> normalize_kite_size
    ; wing_size = None
    ; sesh_type = maybe_string sesh_type |> normalize_sesh_type
    ; board_type = None
    ; foil = None
    ; board = None
    ; location = maybe_string location
    ; comments = maybe_string comments
    }
  | _ -> failwith "parse_2014: unexpected row format"
;;

let parse_2016 row =
  match row with
  | date :: sport :: hours :: lull :: gust :: kite :: sesh_type :: board :: location :: comments :: _ ->
    { date = parse_date date
    ; sport = normalize_sport sport
    ; hours = float_of_string hours
    ; wind_avg = parse_int lull
    ; wind_gust = parse_int gust
    ; kite_size = parse_many kite |> normalize_kite_size
    ; wing_size = None
    ; sesh_type = maybe_string sesh_type |> normalize_sesh_type
    ; board_type = parse_board_types board
    ; foil = None
    ; board = None
    ; location = maybe_string location
    ; comments = maybe_string comments
    }
  | _ -> failwith "parse_2016: unexpected row format"
;;

let parse_2022 row =
  match row with
  | [ _; date; sport; hours; avg; gust; kite; wing; sesh_type; board_type; foil; foil_board; location; comments ] ->
    { date = parse_date date
    ; sport = normalize_sport sport
    ; hours = float_of_string hours
    ; wind_avg = parse_int avg
    ; wind_gust = parse_int gust
    ; kite_size = parse_many kite |> normalize_kite_size
    ; wing_size = parse_many wing
    ; sesh_type = maybe_string sesh_type |> normalize_sesh_type
    ; board_type = parse_board_types board_type
    ; foil = parse_many foil
    ; board = maybe_string foil_board
    ; location = maybe_string location
    ; comments = maybe_string comments
    }
  | _ -> failwith "parse_2022: unexpected row format"
;;

let parse_2024 row =
  match row with
  | [ _; date; sport; hours; avg; gust; kite; wing; board_type; foil; foil_board; location; sesh_type; comments ]
  | [ _; date; sport; hours; avg; gust; kite; wing; board_type; foil; foil_board; location; sesh_type; comments; _ ] ->
    { date = parse_date date
    ; sport = normalize_sport sport
    ; hours = float_of_string hours
    ; wind_avg = parse_int avg
    ; wind_gust = parse_int gust
    ; kite_size = parse_many kite |> normalize_kite_size
    ; wing_size = parse_many wing
    ; sesh_type = maybe_string sesh_type |> normalize_sesh_type
    ; board_type = parse_board_types board_type
    ; foil = parse_many foil
    ; board = maybe_string foil_board
    ; location = maybe_string location
    ; comments = maybe_string comments
    }
  | _ -> failwith "parse_2024: unexpected row format"
;;

let parse_2025 row =
  match row with
  | [ _; date; sport; hours; avg; gust; kite; wing; board_type; foil; board; location; sesh_type; comments ]
  | [ _; date; sport; hours; avg; gust; kite; wing; board_type; foil; board; location; sesh_type; comments; _ ] ->
    { date = parse_date date
    ; sport = normalize_sport sport
    ; hours = float_of_string hours
    ; wind_avg = parse_int avg
    ; wind_gust = parse_int gust
    ; kite_size = parse_many kite |> normalize_kite_size
    ; wing_size = parse_many wing
    ; sesh_type = maybe_string sesh_type |> normalize_sesh_type
    ; board_type = parse_board_types board_type
    ; foil = parse_many foil
    ; board = maybe_string board
    ; location = maybe_string location
    ; comments = maybe_string comments
    }
  | _ -> failwith "parse_2025: unexpected row format"
;;

let parse_file ~schema rows =
  let parser =
    match schema with
    | 2012 -> parse_2012
    | 2013 -> parse_2013
    | 2015 -> parse_2015
    | 2014 -> parse_2014
    | 2016 | 2017 | 2018 | 2019 | 2020 | 2021 -> parse_2016
    | 2022 -> parse_2022
    | 2023 -> parse_2024
    | 2024 -> parse_2024
    | 2025 | 2026 -> parse_2025
    | other -> failwith (Printf.sprintf "Schema for `%d` not defined yet!" other)
  in
  List.map parser rows
;;
